altair/datasets/_readers.py

```diff
@@ -0,0 +1,569 @@
+        """
+        Returns path to datasets cache, if possible.
+
+        Requires opt-in via environment variable::
+
+            Reader._ENV_VAR
+        """
+        if _dir := os.environ.get(self._ENV_VAR):
+            cache_dir = Path(_dir)
+            cache_dir.mkdir(exist_ok=True)
+            return cache_dir
+
+    def _import(self, name: str, /) -> Any:
+        if spec := find_spec(name):
+            return import_module(spec.name)
+        else:
+            reqs = _requirements(self._name)  # type: ignore[call-overload]
+            if isinstance(reqs, tuple):
+                depends = ", ".join(f"{req!r}" for req in reqs) + " packages"
+            else:
+                depends = f"{reqs!r} package"
+
+            msg = (
+                f"Backend {self._name!r} requires the {depends}, but {name!r} could not be found.\n"
+                f"This can be installed with pip using:\n"
+                f"    pip install {name}\n"
+                f"Or with conda using:\n"
+                f"    conda install -c conda-forge {name}"
+            )
+            raise ModuleNotFoundError(msg, name=name)
+
+    def __repr__(self) -> str:
+        return f"Reader[{self._name}]"
+
+    def __init__(self, name: LiteralString, /) -> None: ...
+
+
+class _PandasReader(_Reader["pd.DataFrame", "pd.DataFrame"]):
+    def __init__(self, name: _Pandas, /) -> None:
+        self._name = _requirements(name)
+        if not TYPE_CHECKING:
+            pd = self._import(self._name)
+        self._read_fn = {
+            ".csv": pd.read_csv,
+            ".json": pd.read_json,
+            ".tsv": partial["pd.DataFrame"](pd.read_csv, sep="\t"),
+            ".arrow": pd.read_feather,
+            ".parquet": pd.read_parquet,
+        }
+        self._scan_fn = {".parquet": pd.read_parquet}
+
+
+class _PandasPyArrowReader(_Reader["pd.DataFrame", "pd.DataFrame"]):
+    def __init__(self, name: Literal["pandas[pyarrow]"], /) -> None:
+        _pd, _pa = _requirements(name)
+        self._name = name
+        if not TYPE_CHECKING:
+            pd = self._import(_pd)
+            pa = self._import(_pa)  # noqa: F841
+
+        self._read_fn = {
+            ".csv": partial["pd.DataFrame"](pd.read_csv, dtype_backend=_pa),
+            ".json": partial["pd.DataFrame"](pd.read_json, dtype_backend=_pa),
+            ".tsv": partial["pd.DataFrame"](pd.read_csv, sep="\t", dtype_backend=_pa),
+            ".arrow": partial(pd.read_feather, dtype_backend=_pa),
+            ".parquet": partial(pd.read_parquet, dtype_backend=_pa),
+        }
+        self._scan_fn = {".parquet": partial(pd.read_parquet, dtype_backend=_pa)}
+
+
+def _pl_read_json_roundtrip(source: Path | IOBase, /, **kwds: Any) -> pl.DataFrame:
+    """
+    Try to utilize better date parsing available in `pl.read_csv`_.
+
+    `pl.read_json`_ has few options when compared to `pl.read_csv`_.
+
+    Chaining the two together - *where possible* - is still usually faster than `pandas.read_json`_.
```

Comment:
@joelostblom (https://github.com/vega/altair/pull/3631#issuecomment-2480816377)

> ### Polars
> 
> I am less familiar with polars (although I have been using it more lately and enjoy it!), but your suggestion with round-tripping to csv seems reasonable, unless that is very time consuming (e.g. for the large flights dataset it might take long? or does the lazy loading/executing make this less of an issue?).

So it is faster than `pandas`, but has to be run using **eager** `pl.read_` functions instead of **lazy** `pl.scan_` functions